statie/source/_posts/2019/2019-05-06-single-responsibility-principle-back-to-basics-or-over-engineering.md

```diff
@@ -0,0 +1,331 @@
+
+Mixed responsibility also makes usage more difficult.
+If we want to use JSON as the class provides, it is usable.
+But once we need a different JSON format, we have to change the serialization method first.
+And the class becomes more and more complex.
+
+To be honest there may be use-cases for the internal JSON serialization.
+If an object is serialized to JSON heavily in the system, then the internal serialization could make sense as it is a common use-case.
+But it is not the money case, at least not in this system.
+
+### External serialization
+```php
+class JsonMoneySerializer {
+    public function serialize(Money $money): string {
+        return [
+            'amount' => $money->getAmount(),
+        ];
+        return json_encode($array);
+    }
+}
+```
+Usage
+```php
+$stringJson = $this->jsonMoneySerializer->serialize($money);
+```
+Serializer has a clear responsibility - serializing money to JSON.
+Money class is not changed and is not messed by the serialization code. 
+
+When we'll need a different JSON structure, we'll create a new serializer.
+We won't have to touch the current code, so we're less likely to break it.
+We end up with short, even dumb, code that is easy to understand and combine.
+
+## Price module
+
+We have a module that is responsible for the price.
+It answers questions like
+*What is the price of the product?*
+*What is the price of the shipping?*
+*What is the overall price of the order?*
+
+Now we have a new use case *Sort products by price*.
+Should this behavior be a part of the pricing module?
+Well maybe yes and maybe no.
+
+A good practice I found is to put this new responsibility into a separate module that uses the price module.
+As we understand more, we may find that the separation was a good idea because the new module is at the end responsible for preparing data for the web,
+caching, and that is a clear separated responsibility.
+
+If we add this new responsibility into an existing module, it will start growing.
+*Where should I put price caching?* Into the price module, it deals with price.
+*Where should I put preparing price data for web?* Into the price module, it still fits there.
+And we end up with one huge module that has multiple responsibilities, is difficult to understand and nearly impossible to maintain as all pieces depend on each other.
+
+On the other hand, we may later find that this new module was never used for anything else than sorting and that it would be more meaningful
+to have the functionality within the price module.
+In such a case, it is a good idea to merge these two modules into one with extended responsibility.
+
+The trick is that merging two modules into one is often much easier than splitting one module into two.
+So generally I suggest starting with a separated module for a responsibility that seems to not fit into an existing module
+while keeping an option to merge them later.
```

Comment:
There is a long section without headline. It would make text easier to scan if there would be one